File: solutions/python/cricinfo/match_service.py
from match import Match
from match_status import MatchStatus
from threading import Lock

class MatchService:
    _instance = None
    _lock = Lock()

    def __init__(self):
        self.matches = {}

    @classmethod
    def get_instance(cls):
        with cls._lock:
            if cls._instance is None:
                cls._instance = cls()
        return cls._instance

    def add_match(self, match):
        self.matches[match.get_id()] = match

    def get_match(self, match_id):
        return self.matches.get(match_id)

    def get_all_matches(self):
        return list(self.matches.values())

    def update_match_status(self, match_id, status):
        match = self.get_match(match_id)
        if match:
            match.set_status(status)

File: solutions/golang/cricinfo/match_service.go
package cricinfo

import (
	"sync"
)

type MatchService struct {
	matches map[string]*Match
	mu      sync.RWMutex
}

var (
	matchServiceInstance *MatchService
	matchServiceOnce     sync.Once
)

func GetMatchService() *MatchService {
	matchServiceOnce.Do(func() {
		matchServiceInstance = &MatchService{
			matches: make(map[string]*Match),
		}
	})
	return matchServiceInstance
}

func (ms *MatchService) AddMatch(match *Match) {
	ms.mu.Lock()
	defer ms.mu.Unlock()
	ms.matches[match.ID] = match
}

func (ms *MatchService) GetMatch(matchID string) *Match {
	ms.mu.RLock()
	defer ms.mu.RUnlock()
	return ms.matches[matchID]
}

func (ms *MatchService) GetAllMatches() []*Match {
	ms.mu.RLock()
	defer ms.mu.RUnlock()

	matches := make([]*Match, 0, len(ms.matches))
	for _, match := range ms.matches {
		matches = append(matches, match)
	}
	return matches
}

func (ms *MatchService) UpdateMatchStatus(matchID string, status MatchStatus) {
	ms.mu.Lock()
	defer ms.mu.Unlock()
	if match, exists := ms.matches[matchID]; exists {
		match.Status = status
	}
}
